internal/controller/apiextensions/composite/composed.go

```diff
@@ -238,3 +278,324 @@ func IsReady(_ context.Context, cd resource.Composed, t v1.ComposedTemplate) (bo
+	if len(templates) == 0 {
+		return r.anonymous.Compose(ctx, cr, comp)
+	}
+
+	resources := map[string]resource.Composed{}
+	for _, ref := range cr.GetResourceReferences() {
+		cd := composed.New(composed.FromReference(ref))
+		nn := types.NamespacedName{Namespace: ref.Namespace, Name: ref.Name}
+		if err := r.client.Get(ctx, nn, cd); resource.IgnoreNotFound(err) != nil {
+			return CompositionResult{}, errors.Wrap(err, errGetComposed)
+		}
+
+		name := GetCompositionTemplateName(cd)
+		if meta.WasCreated(cd) && name == "" {
+			// All templates are named but this composed resource isn't. We are
+			// most likely updating from a Composition with anonymous templates
+			// to one with named templates - we need to run the anonymous
+			// composer to propagate template names to the existing composed
+			// resources.
+			return r.anonymous.Compose(ctx, cr, comp)
+		}
+
+		resources[name] = cd
+	}
+
+	// We ensure all resources can be rendered before we apply any.
+	refs := make([]corev1.ObjectReference, len(comp.Spec.Resources))
+	for i, tpl := range comp.Spec.Resources {
+		cd := composed.New()
+		if existing, ok := resources[*tpl.TemplateName]; ok {
+			cd.SetNamespace(existing.GetNamespace())
+			cd.SetName(existing.GetName())
+		}
+
+		if err := r.composed.Render(ctx, cr, cd, tpl); err != nil {
+			return CompositionResult{}, errors.Wrapf(err, errFmtRenderName, *tpl.TemplateName)
+		}
+
+		resources[*tpl.TemplateName] = cd
+		refs[i] = *meta.ReferenceTo(cd, cd.GetObjectKind().GroupVersionKind())
+	}
+
+	for name, cd := range resources {
+		if _, ok := templates[name]; !ok {
+			// This composed resource exists, but the template that was used to
+			// create it does not. It should be garbage collected.
+			if err := r.client.Delete(ctx, cd); resource.IgnoreNotFound(err) != nil {
+				return CompositionResult{}, errors.Wrap(err, errGCComposed)
+			}
+			continue
+		}
+	}
+
+	// We need to persist our resource references after garbage collection has
+	// succeeded but before we create any new composed resources. If we persist
+	// references before garbage collection succeeds we couldn't recover from an
+	// error during garbage collection; on the next reconcile we'd have orphaned
+	// resources that weren't in our references array. If we don't persist
+	// references before we try to create new resources an error during creation
+	// may cause the next reconcile to generate new names and thus create
+	// duplicate resources.
+	cr.SetResourceReferences(refs)
+	if err := r.client.Update(ctx, cr); err != nil {
+		return CompositionResult{}, errors.Wrap(err, errUpdate)
+	}
+
+	// The resources map should now contain resources that were rendered using
+	// the current array of templates, as well as any existing resources that no
+	// longer have a corresponding template. The former should be applied, while
+	// the latter should be deleted.
+	conn := managed.ConnectionDetails{}
+	ready := 0
+	for name, cd := range resources {
```

Comment:
Iterating over the map of resources will result in a different order each reconciliation. In practice, this might be low impact, but seems like a difference that was noteworthy.

This may allow an XR resource to work around a stuck resource after a few reconcile loops, as the composed resource introducing the error may fall in a different position when iterating on the map. But if applying all composed resources besides the ones which experience an error is desirable, it is probably worth doing so explicitly with continuation and an error array, rather than using the reconciler back-off and randomness.

Reply:
Good observation. That said, I'm not overly concerned about optimising for either applying managed resources in the order of their templates or making the most progress possible in a single reconcile. Arbitrary order of operation and retrying when we can't do something both seem fairly in line with the eventually consistent nature of the Kubernetes API to me.